### ISSUE TITLE
Document disambiguation strategy when a token's prefix matches another token

### ISSUE THREAD

**Issue by marijnh** [maciejhirsz/logos#18]
(Very cool library!)

From experimenting it looks like it prefers the longest match? That's probably a reasonable strategy. But when reading the readme I was a bit confused about how it'd handle matching of (from the example) `"fast"` versus `Text` tokens in input like `"faster"`, so it'd help to make that explicit.

**Comment by maciejhirsz** [comment]
Yeah, `"faster"` would produce a `Text` in the readme example. There are probably more hidden assumptions made that need documenting.

**Comment by maciejhirsz** [comment]
@toshokan no, nothing should have changed. I'll investigate, but that sounds like a bug.

**Comment by maciejhirsz** [comment]
@toshokan I can't seem to be able to reproduce it. The readme example is being tested in CI (as a doc comment code snippet).

**Comment by toshokan** [comment]
The readme example itself does work fine on either version.
```rust
use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
enum Token {
    #[end]
    End,
    #[error]
    Error,
    #[token = "fast"]
    Fast,
    #[regex = "[a-zA-Z]+"]
    Text,
}

fn main() {
    let mut lexer = Token::lexer("faster");
    assert_eq!(lexer.token, Token::Text);
    assert_eq!(lexer.slice(), "faster");
    assert_eq!(lexer.range(), 0..6);
    lexer.advance();
    assert_eq!(lexer.token, Token::End);
}
```
This works and passes on `0.9.7` but fails (the first token is `Token::Fast`) on `0.10.0-rc1`

**Comment by maciejhirsz** [comment]
Cheers, I'll work on fixing that before getting 0.10 to release state. My hunch is that it's due to ordering of the patterns.

**Comment by strax** [comment]
Curiously it seems that when a token is at most three characters, it will be eagerly matched even if there is a longest match (both 0.9.7 and 0.10.0-rc2). Adapting @toshokan's example:

```rust
use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
enum Token {
    #[end]
    End,
    #[error]
    Error,
    #[token = "fas"]
    Fas,
    #[regex = "[a-zA-Z]+"]
    Text,
}

fn main() {
    let mut lexer = Token::lexer("faster");
    assert_eq!(lexer.token, Token::Text);
    assert_eq!(lexer.slice(), "faster");
    lexer.advance();
    assert_eq!(lexer.token, Token::End);
}
```

As mentioned, the token "fast" will produce correct lexemes on 0.9.7, but "fas" will incorrectly produce a `Token::Fas`.

**Comment by maciejhirsz** [comment]
This should be, at very long last, fixed in 0.10.0-rc3 with the new refactor (#94). Documentation for disambiguation will follow before the final release of 0.10, TL;DR for it right now it's done by comparing priority, where priority is calculated as longest byte-sequence matched _without_ repeats:

+ `[a-z]*` has priority 0.
+ `[a-z]+` has priority 1.
+ `foobar` has priority 6.
+ `(fooz|baz)+qux` has priority 7.

I think I'll double priority for single-byte patterns, so that more specialized `f` will have precedence over more general `[a-z]+`.
